lxd/db/update.go

```diff
@@ -29,66 +30,43 @@ import (
    Only append to the updates list, never remove entries and never re-order them.
 */
 
-var dbUpdates = []dbUpdate{
-	{version: 1, run: dbUpdateFromV0},
-	{version: 2, run: dbUpdateFromV1},
-	{version: 3, run: dbUpdateFromV2},
-	{version: 4, run: dbUpdateFromV3},
-	{version: 5, run: dbUpdateFromV4},
-	{version: 6, run: dbUpdateFromV5},
-	{version: 7, run: dbUpdateFromV6},
-	{version: 8, run: dbUpdateFromV7},
-	{version: 9, run: dbUpdateFromV8},
-	{version: 10, run: dbUpdateFromV9},
-	{version: 11, run: dbUpdateFromV10},
-	{version: 12, run: dbUpdateFromV11},
-	{version: 13, run: dbUpdateFromV12},
-	{version: 14, run: dbUpdateFromV13},
-	{version: 15, run: dbUpdateFromV14},
-	{version: 16, run: dbUpdateFromV15},
-	{version: 17, run: dbUpdateFromV16},
-	{version: 18, run: dbUpdateFromV17},
-	{version: 19, run: dbUpdateFromV18},
-	{version: 20, run: dbUpdateFromV19},
-	{version: 21, run: dbUpdateFromV20},
-	{version: 22, run: dbUpdateFromV21},
-	{version: 23, run: dbUpdateFromV22},
-	{version: 24, run: dbUpdateFromV23},
-	{version: 25, run: dbUpdateFromV24},
-	{version: 26, run: dbUpdateFromV25},
-	{version: 27, run: dbUpdateFromV26},
-	{version: 28, run: dbUpdateFromV27},
-	{version: 29, run: dbUpdateFromV28},
-	{version: 30, run: dbUpdateFromV29},
-	{version: 31, run: dbUpdateFromV30},
-	{version: 32, run: dbUpdateFromV31},
-	{version: 33, run: dbUpdateFromV32},
-	{version: 34, run: dbUpdateFromV33},
-	{version: 35, run: dbUpdateFromV34},
-	{version: 36, run: dbUpdateFromV35},
-}
-
-type dbUpdate struct {
-	version int
-	run     func(previousVersion int, version int, db *sql.DB) error
-}
-
-func (u *dbUpdate) apply(currentVersion int, db *sql.DB) error {
-	// Get the current schema version
-
-	logger.Debugf("Updating DB schema from %d to %d", currentVersion, u.version)
-
-	err := u.run(currentVersion, u.version, db)
-	if err != nil {
-		return err
-	}
-
-	_, err = db.Exec("INSERT INTO schema (version, updated_at) VALUES (?, strftime(\"%s\"));", u.version)
-	if err != nil {
-		return err
-	}
-
-	return nil
+var updates = []schema.Update{
+	updateFromV0,
+	updateFromV1,
+	updateFromV2,
+	updateFromV3,
+	updateFromV4,
+	updateFromV5,
+	updateFromV6,
+	updateFromV7,
+	updateFromV8,
+	updateFromV9,
+	updateFromV10,
+	updateFromV11,
+	updateFromV12,
+	updateFromV13,
+	updateFromV14,
+	updateFromV15,
+	updateFromV16,
+	updateFromV17,
+	updateFromV18,
+	updateFromV19,
+	updateFromV20,
+	updateFromV21,
+	updateFromV22,
+	updateFromV23,
+	updateFromV24,
+	updateFromV25,
+	updateFromV26,
+	updateFromV27,
+	updateFromV28,
+	updateFromV29,
+	updateFromV30,
+	updateFromV31,
+	updateFromV32,
+	updateFromV33,
+	updateFromV34,
+	updateFromV35,
```

Comment:
So I'm not particularly fond of effectively hiding the target schema version.

It makes it harder for a contributor to figure out exactly what schema versions are, though the function names would probably still make it reasonably easy to find (so long as they think of looking at the version prior to their current one).

What worries me is how easy it now is to just add a couple of entries to the updates list without fully understanding that this is going to bump the schema version. You then remove one you added by mistake and you now have an invalid schema version, or worse, you re-order two lines and now you potentially broke your database.

While all of this was possible before, the fact that the schema version had to be manually put up next to the update function was making it much much harder for someone to screw up.

To be clear, this is an argument for contributors doing local changes to their LXD before submitting a branch upstream. Schema updates aren't particularly easy to do and even less to iterate on locally. So keeping things visually clear is pretty important here so that if they do mess something up, they can figure out what to revert in their database schema, what to mangle in the schema table and can then run the upgrade again.

Reply:
In the current code there isn't two source of truth. The version listed in the list is the schema version that will be inserted into the database upon success of the run function.

```
var dbUpdates = []dbUpdate{
	{version: 1, run: foo},
	{version: 3, run: bar},
	{version: 2, run: egg},
}
```

Makes it clear that you'll first run "foo" which will cause the schema row "1" to be inserted, then run "bar" which will cause the schema row "3" to be inserted and lastly run "egg" which will cause "2" to be inserted.

The upgrade would obviously fail if egg was to depend on the schema change from bar, but it's visibly wrong and you can trivially see what function is tied to what schema version, then can manually revert that change in your database before fixing your code.


Another case I just thought about where declaring the schema version for every db update function may save us some significant headache is with cherry-picks. It's not uncommon for git cherry-pick to think it managed to cleanly apply a commit but have it applied in the wrong place in a list.

With the current approach this is either harmless or causes a duplicate schema version in the list which will trigger an error. With the new approach, this would inject a new entry in the middle of the list, therefore re-numbering the schema versions for all entries after it.

Reply:
The thing with:

```
var dbUpdates = []dbUpdate{
	{version: 1, run: foo},
	{version: 3, run: bar},
	{version: 2, run: egg},
}
```

is that the current code will silently ignore the last update ("egg'):

```go
	for _, update := range dbUpdates {
		if update.version <= currentVersion {
			continue
		}
                // ...
		currentVersion = update.version
	}
```

and this essentially happens because there authoritative source of truth for the version is the version field, and the authoritative source of truth for the *order* of application is the position in the slice. Those two bits of information can clash at any time, so you either live with the fact that the model lets you express inconsistencies or you add code that double checks those inconsistencies.

I'd propose to use ```map[schema.Version]schema.Update``` (iow ```map[int]func(tx *sql.Tx) error```) which forces the source of truth of the order to be the version as well.

Reply:
Ok, lets do that.

Reply:
Done.